### PR TITLE
Pick and choose numpy version based on Python 2 or 3.

### LINKED ISSUE #27435: Cannot install pandas 0.24.2 from source with Python 2.7 into an environment without NumPy
#### Code Sample, a copy-pastable example if possible

```bash
RuntimeError: Python version >= 3.5 required.
```
#### Problem description

Under this docs: https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.24.0.html

0.24.x is supposed to be backward compatible to Python 2.

```
The 0.24.x series of releases will be the last to support Python 2. Future feature releases will support Python 3 only. See Plan for dropping Python 2.7 for more details
```

#### Expected Output

0.24.2 should still work on Python 2.

### PR DESCRIPTION
**Pull request by didip**
- [x] closes #27435
- [ ] tests added / passed
- [ ] passes `black pandas` (not applicable, `black setup.py` changes too many quotes, polluting the diff).
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

cc @jorisvandenbossche

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Thanks for the PR!

Hmm, there a couple of failures. Not directly related to your change, but related to the fact that we didn't run those tests (at the state of 0.24.2) for some time, and fixes have been made for changed dependencies. 
See eg https://github.com/pandas-dev/pandas/issues/26725, https://github.com/pandas-dev/pandas/pull/26726

There also seem to be some actual failures with numpy 1.17, so an option could be to restrict it to lower versions also for python 3 ?

**Comment by didip** [comment]
Thanks for the quick response.

Do you have documented versions of what the maximum versions pandas should have on Python 3?

**Comment by TomAugspurger** [comment]
I haven't looked at the failures, but are they actual, noticeable bugs, or
just failures in the test?

If they're real issues then we can just have pandas 0.24.x require NumPy
<1.17 across the board.

**Comment by TomAugspurger** [comment]
Yeah, numpydev can certainly be ignored.

For the rest, perhaps try adding back the `free` channel in each
environment that needs it?

File: setup.py [hunk 1]
```diff
@@ -31,13 +31,22 @@ def is_platform_mac():
...
 
 min_numpy_ver = '1.12.0'
+max_numpy_ver = ''
+
+if (sys.version_info < (3, 0)):
+    max_numpy_ver = '1.17'
+
+numpy_install_rule = 'numpy >= {numpy_ver}'.format(numpy_ver=min_numpy_ver)
+if max_numpy_ver:
+    numpy_install_rule += ', < {numpy_ver}'.format(numpy_ver=max_numpy_ver)
```
Comment (datapythonista) [inline setup.py]: Just a style preference, but what fo you think about the less verbose option:

```python
numpy_requires = 'numpy >= {min_numpy_ver}{max_numpy_rule}'.format(
    min_numpy_ver=min_numpy_ver,
    max_numpy_rule=', < {}'.format(max_numpy_ver_py2) if sys.version_info[0] < 3 else '')
```

**Review by datapythonista — COMMENTED** [review]
Thanks for the fixing this @didip. Can you remove the tailing whitespace that is breaking the CI, and check the styling comment?

**Comment by datapythonista** [comment]
Also, if you can please merge master into this branch.

**Comment by TomAugspurger** [comment]
I think this is targeting 0.24.x directly, since we don't want these changes on master.

**Comment by didip** [comment]
Agree with @TomAugspurger

**Comment by TomAugspurger** [comment]
We still need the linting issues fixed though.

**Comment by didip** [comment]
Yes, I’ll try to find sometime today to address it.

**Comment by didip** [comment]
Where is the CI/CD output that shows failed linting?

**Comment by datapythonista** [comment]
yes, but the build is too old and is not available anymore. You can run `./ci/code_checks.sh` and should report where are the tailing spaces.

**Comment by didip** [comment]
There are tons of lint problems when running `./ci/code_checks.sh` but I don't see one that pertains to `setup.py`.

**Comment by datapythonista** [comment]
Not quite sure what's the problem with the docstrings, but you've also got this error in the CI `pytables needs a release beyong 3.4.4 to support numpy 1.16x`.

Do you mind having a look @didip. We can't merge this until the CI is green.

**Comment by WillAyd** [comment]
At this point we don't have any further plans for a 24.x release, so unfortunately I don't think this is something we will address (if anyone disagrees feel free to reopen)

**Comment by jorisvandenbossche** [comment]
As far as I recall, we said to give the opportunity to interested contributors, so let's wait a bit more to see if @didip has time to update this.

**Comment by jbrockmendel** [comment]
@didip it looks like some xfailed pytables tests need to be updated for the changes in this PR to work.  can you take a look and update?

**Comment by didip** [comment]
Upgraded `pytables` to 3.5.2. `sh test_fast.sh` now actually runs end-to-end with 20 failures.

**Comment by didip** [comment]
`sh test.sh` ran until the end with 45 failures.

Anything else I need to do? Is this good enough to merge?

**Comment by jbrockmendel** [comment]
> sh test.sh ran until the end with 45 failures.  Anything else I need to do? Is this good enough to merge?

You'll need to get the tests passing.  This can include xfailing if appropriate.

**Comment by didip** [comment]
@jbrockmendel They look unrelated though. Example:

```
XFAIL pandas/tests/test_strings.py::TestStringMethods::test_api_per_method[empty0-rsplit-Index-category]
  reason: Raising too restrictively; solved by GH 23167
```

**Comment by didip** [comment]
Ran `sh ci/code_checks.sh`. All of its failures are DocTest.

All of them failed because the expectation and results are in the wrong order. Examples:

```
4253 Examples
4254 --------
4255 >>> df = pd.DataFrame({'name': ['Raphael', 'Donatello'],
4256 ...                    'mask': ['red', 'purple'],
4257 ...                    'weapon': ['sai', 'bo staff']})
4258 >>> df.to_csv(index=False)
Expected:
    'name,mask,weapon\nRaphael,red,sai\nDonatello,purple,bo staff\n'
Got:
    'mask,name,weapon\nred,Raphael,sai\npurple,Donatello,bo staff\n'
```

or another one:

```
2665         >>> d2 = {'name': ['Alice', 'Bob'],
2666         ...       'score': [9.5, 8],
2667         ...       'employed': [False, True],
2668         ...       'kids': [0, 0]}
2669         >>> df2 = pd.DataFrame(data=d2)
2670         >>> df2
Differences (unified diff with -expected +actual):
    @@ -1,3 +1,3 @@
    -    name  score  employed  kids
    -0  Alice    9.5     False     0
    -1    Bob    8.0      True     0
    +   employed  kids   name  score
    +0     False     0  Alice    9.5
    +1      True     0    Bob    8.0
```

File: requirements-dev.txt [hunk 2]
```diff
@@ -31,7 +31,7 @@ nbsphinx
 numexpr>=2.6.8
 openpyxl
 pyarrow>=0.9.0
-tables>=3.4.2
+tables>=3.5.2
```
Comment (jreback) [inline requirements-dev.txt]: i don't think this would work because it would require a newer numpy than is allowed in this version (1.13)

**Review by jreback — CHANGES_REQUESTED** [review]
@didip happy to take this, but needs to pass the CI (and I know you tried to fix)

ping if you can get this too work.

**Comment by alimcmaster1** [comment]
@didip - thanks for the effort so far - do you still want to work on this?

**Comment by didip** [comment]
To be honest, I am running out of capacity to continue this work. I am completely unfamiliar with the build system to properly debug further problems.

**Comment by jreback** [comment]
closing, but still happy for someone to push a PR to close this issue.

**Pull request closed without merging**